tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3090,8 +3100,15 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
   int64 num_threads_y = reduction_dimensions.is_row_reduction ? 1 : kWarpSize;
   int64 num_threads_x = [&] {
     if (reduction_dimensions.is_row_reduction) {
+      int cc_major = 0, cc_minor = 0;
+      ir_emitter_context_->device_description().cuda_compute_capability(
+          &cc_major, &cc_minor);
+      int64 num_threads_x = kWarpSize * kWarpSize;
+      if (cc_major >= 6 && smallest_input_dtype_bits <= 16) {
+        num_threads_x = kWarpSize * 8;
```

Comment:
Sorry, this gets a bit complicated here: such a change would require a change here as well: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/tree_reduction_rewriter.cc#L43, which would then require changes to the corresponding tests.

Could we just commit a tiling change here, and then have a separate PR with a discussion for changing the number of threads in a block?

Reply:
I can't. For float16, it would works, I would get a 2.5x speed up instead of 3.5x.
But for u8, I get a 0.95 speed up. So this would bring a regression.
I'll try to update the corresponding code and refactor to have only 1 version.

Reply:
I tried it rapidly. This generate 5 tests failure. They fail because the tests expect 2 reductions, but now there is 4 reductions.
For this case, we want the biggest blocks as otherwise it would slow down the deterministic reduction and we do not want that.
So I could add an attribute inside the Reduce operations called "biggest_block" to prevent this slowdown and would fix the test failure at the same time.

What do you think of that solution? I already added an attribute to an XLA operation. If you agree with that solution, I'll implement it.

Reply:
> I can't. For float16, it would works, I would get a 2.5x speed up instead of 3.5x.
But for u8, I get a 0.95 speed up

Could we still commit this? 2.5x speed up for fp16 is very impressive, and it sounds well worth the trade-off for 0.95x u8 slowdown.

Also do you know why changing the tiling without cutting down the number of threads creates the slowdown?

> So I could add an attribute inside the Reduce operations called "biggest_block" to prevent this slowdown

Reply:
>So I could add an attribute inside the Reduce operations called "biggest_block" to prevent this slowdown

I am confused, what would this attribute capture?

Reply:
If we unroll more per threads, without lowering the number of threads per blocks, it mean we will launch less blocks. For small cases, we won't have enough block to saturate the IO and each block will take longer. So this is normal for small cases to be slowed down.
If you do not mind the 0.95x slowdown for u8, I'll remove that from this PR.

About "biggest_block". The idea is to tell the Reduce code emitter some information about how it must implement the code. There could be a normal case and a case where it will try to make each block read more elements. You want the deterministic version to use that case.
We could name it "deterministic" and if it can't use enough threads, then it could error.
We could even have this mode query the max number of threads per blocks, so it could have more then 1024 threads per blocks.